pkg/server/plugin/nodeattestor/tpmdevid/devid.go

```diff
@@ -0,0 +1,491 @@
+		if !oid.Equal(subjectAlternativeNameOID) {
+			unhandledExtensions = append(unhandledExtensions, oid)
+		}
+	}
+
+	ekCert.UnhandledCriticalExtensions = unhandledExtensions
+
+	_, err := ekCert.Verify(x509.VerifyOptions{
+		KeyUsages: []x509.ExtKeyUsage{x509.ExtKeyUsageAny},
+		Roots:     roots,
+	})
+	if err != nil {
+		return fmt.Errorf("endorsement certificate verification failed: %w", err)
+	}
+
+	return nil
+}
+
+// verifyEKsMatch checks that the public key generated using the EK template
+// matches the public key included in the Endorsement Certificate.
+func verifyEKsMatch(ekCert *x509.Certificate, ekPub tpm2.Public) error {
+	keyFromCert, ok := ekCert.PublicKey.(*rsa.PublicKey)
+	if !ok {
+		return fmt.Errorf("key from certificate is not an RSA key")
+	}
+
+	cryptoKey, err := ekPub.Key()
+	if err != nil {
+		return fmt.Errorf("cannot get template key: %w", err)
+	}
+
+	keyFromTemplate, ok := cryptoKey.(*rsa.PublicKey)
+	if !ok {
+		return fmt.Errorf("key from template is not an RSA key")
+	}
+
+	if keyFromCert.E != keyFromTemplate.E {
+		return fmt.Errorf("exponent missmatch")
+	}
+
+	if keyFromCert.N.Cmp(keyFromTemplate.N) != 0 {
+		return fmt.Errorf("modulus missmatch")
+	}
+
+	return nil
+}
+
+func verifyDevIDCertification(pubAK, pubDevID *tpm2.Public, attestData, attestSig []byte) error {
+	err := checkSignature(pubAK, attestData, attestSig)
+	if err != nil {
+		return err
+	}
+
+	data, err := tpm2.DecodeAttestationData(attestData)
+	if err != nil {
+		return err
+	}
+
+	if data.AttestedCertifyInfo == nil {
+		return errors.New("missing certify info")
+	}
+
+	ok, err := data.AttestedCertifyInfo.Name.MatchesPublic(*pubDevID)
+	if err != nil {
+		return err
+	}
+
+	if !ok {
+		return errors.New("certify failed")
+	}
+
+	return nil
+}
+
+func checkSignature(pub *tpm2.Public, data, sig []byte) error {
+	key, err := pub.Key()
+	if err != nil {
+		return err
+	}
+
+	rsaKey, ok := key.(*rsa.PublicKey)
+	if !ok {
+		return errors.New("only RSA keys are supported")
+	}
+
+	sigScheme, err := getSignatureScheme(*pub)
+	if err != nil {
+		return err
+	}
+
+	hash, err := sigScheme.Hash.Hash()
+	if err != nil {
+		return err
+	}
+
+	h := hash.New()
+	_, err = h.Write(data)
+	if err != nil {
+		return err
+	}
+
+	hashed := h.Sum(nil)
+
+	return rsa.VerifyPKCS1v15(rsaKey, hash, hashed, sig)
+}
+
+func getSignatureScheme(pub tpm2.Public) (*tpm2.SigScheme, error) {
+	canSign := (pub.Attributes & tpm2.FlagSign) == tpm2.FlagSign
+	if !canSign {
+		return nil, errors.New("not a signing key")
+	}
+
+	switch pub.Type {
+	case tpm2.AlgRSA:
+		params := pub.RSAParameters
+		if params == nil {
+			return nil, errors.New("malformed key")
+		}
+
+		return params.Sign, nil
+
+	case tpm2.AlgECDSA:
+		params := pub.ECCParameters
+		if params == nil {
+			return nil, errors.New("malformed key")
+		}
+
+		return params.Sign, nil
+
+	default:
+		return nil, fmt.Errorf("unsupported key type 0x%04x", pub.Type)
```

Comment:
`0x%04x` is always returned instead of `pub.Type`.

Reply:
Hmm I think it is working ok? https://play.golang.org/p/2ecIQALLuu7 but maybe I'm missing something

Reply:
You are right!